networks/2_get_network_communities.py

```diff
@@ -0,0 +1,127 @@
+from core_acc_modules import paths
+
+# ## Set user parameters
+#
+# For now we will vary the correlation threshold (`corr_threshold`) but keep the other parameters consistent
+#
+# We will run this notebook for each threshold parameter
+
+# +
+# Params
+corr_threshold = 0.5
+
+# Correlation matrix files
+pao1_corr_filename = os.path.join(
+    paths.LOCAL_DATA_DIR, f"pao1_corr_{corr_threshold}.tsv"
+)
+pa14_corr_filename = os.path.join(
+    paths.LOCAL_DATA_DIR, f"pa14_corr_{corr_threshold}.tsv"
+)
+# -
+
+# Load correlation data
+pao1_corr = pd.read_csv(pao1_corr_filename, sep="\t", index_col=0, header=0)
+pa14_corr = pd.read_csv(pa14_corr_filename, sep="\t", index_col=0, header=0)
+
+# ## Module detection
+# To detect modules, we will use a clustering algorithm
+
+# ### DBSCAN
+# [DBSCAN](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.DBSCAN.html#sklearn.cluster.DBSCAN):  Density-Based Spatial Clustering of Applications with Noise views clusters as areas of high density separated by areas of low density. The central component to the DBSCAN is the concept of _core samples_, which are samples that are in areas of high density. A cluster is therefore a set of _core samples_ that are close to each other (measured by some distance measure) and a set of non-core samples that are close to a core sample (but are not themselves core samples).
+#
+# A cluster is a set of core samples that can be built by recursively taking a core sample, finding all of its neighbors that are core samples, finding all of their neighbors that are core samples, and so on. A cluster also has a set of non-core samples, which are samples that are neighbors of a core sample in the cluster but are not themselves core samples. Intuitively, these samples are on the fringes of a cluster.
+#
+# * We define a core sample as being a sample in the dataset such that there exist `min_samples` other samples within a distance of `eps`, which are defined as neighbors of the core sample.
+# * Using all default values
+#
+
+# +
+# Clustering using DBSCAN
+# pao1_clustering = DBSCAN().fit(pao1_corr)
+# pa14_clustering = DBSCAN().fit(pa14_corr)
+# -
+
+# ### Hierarchal clustering
+# [Hierachal clustering](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.AgglomerativeClustering.html#sklearn.cluster.AgglomerativeClustering): Initially, each object is assigned to its own cluster and then the algorithm proceeds iteratively, at each stage joining the two most similar clusters (i.e. linkage distance is minimized), continuing until there is just a single cluster.
+#
+# * n_cluster: The number of clusters to find.
+# * linkage: Criterion used to determine distance between observations. 'average'=average distance of each observation in the two sets.
+#
+# * Note: It looks like this method tends to produce 1 very large cluster. To break this up we will iteratively apply hierarchal clustering on the largest cluster.
```

Comment:
You could also try changing `n_clusters` - setting this to a larger number might give you a more even size distribution.

Reply:
This parameter confused me a bit. I wasn't sure if `n_clusters` referred to 1) the total number of final clusters or 2) if it was the number of clusters per merge (which I assumed was 2, but maybe that's not the case). I guess if it were (1) then I felt a little hesitant setting a number of total clusters to expect. One thing I was considering applying was this hdbscan that has fewer params to get around this too.

Reply:
I think it's the total number of final clusters (i.e. where to cut the dendrogram to give the final output clusters).

I don't feel strongly about trying it, though - if you'd rather just run it recursively or use another method that's totally fine. You could also chat with Milton if you're trying to decide on a method, I know he's done a lot of clustering stuff in the past so he might know of some options that I'm not aware of.